test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -35,117 +36,126 @@ const (
-		return fmt.Errorf("failed to start registrar: %v", err)
+	d, err := kubeletplugin.Start(ex.logger, driverName, endpoint, draAddress, pluginRegistrationPath, ex)
+	if err != nil {
+		return fmt.Errorf("start kubelet plugin: %v", err)
 	}
+	ex.d = d
 
 	return nil
+}
 
+// stop ensures that all servers are stopped and resources freed.
+func (ex *examplePlugin) stop() {
+	ex.d.Stop()
 }
 
+// NodePrepareResource ensures that the CDI file for the claim exists. It uses
+// a deterministic name to simplify NodeUnprepareResource (no need to remember
+// or discover the name) and idempotency (when called again, the file simply
+// gets written again).
 func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
-
+	logger := klog.FromContext(ctx)
 	kind := ex.driverName + "/" + deviceName
+	filePath := ex.getJSONFilePath(req.ClaimUid)
 
-	klog.Infof("NodePrepareResource is called: request: %+v", req)
-
-	klog.Infof("Creating CDI File")
-	// make resourceHandle in the form of map[string]string
-	var decodedResourceHandle map[string]string
-	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	// Determine environment variables.
+	var env map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &env); err != nil {
+		return nil, fmt.Errorf("unmarshal resource handle: %v", err)
 	}
 
-	// create ContainerEdits in device spec and ContainerEdits
+	// CDI wants env variables as set of strings.
 	envs := []string{}
-	for key, val := range decodedResourceHandle {
+	for key, val := range env {
 		envs = append(envs, key+"="+val)
 	}
-	// create spec
+
 	spec := specs.Spec{
 		Version: cdiVersion,
 		Kind:    kind,
-		Devices: []specs.Device{{
-			Name: deviceName,
-			ContainerEdits: specs.ContainerEdits{
-				Env: envs,
+		Devices: []specs.Device{
+			{
+				Name: deviceName,
+				ContainerEdits: specs.ContainerEdits{
+					Env: envs,
+				},
 			},
-		}},
-		ContainerEdits: specs.ContainerEdits{},
+		},
 	}
-	cdiSpec, err := cdiapi.NewSpec(&spec, ex.getJsonFilePath(req.ClaimUid), 1)
+	cdiSpec, err := cdiapi.NewSpec(&spec, filePath, 1)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create new spec: %v", err)
+		return nil, fmt.Errorf("create spec: %v", err)
 	}
-	klog.V(5).InfoS("CDI spec is validated")
 
-	// create bytes from spec, which would be written into the json file
-	jsonBytes, err := json.Marshal(*cdiSpec)
+	buffer, err := json.Marshal(*cdiSpec)
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal CDI spec: %v", err)
+		return nil, fmt.Errorf("marshal spec: %v", err)
 	}
 
-	// create json file
-	filePath := ex.getJsonFilePath(req.ClaimUid)
-	if err := os.WriteFile(filePath, jsonBytes, os.FileMode(0644)); err != nil {
+	if err := os.WriteFile(filePath, buffer, os.FileMode(0644)); err != nil {
```

Comment:
Have you considered using [this API](https://github.com/container-orchestrated-devices/container-device-interface/blob/v0.5.0/pkg/cdi/spec.go#L116) ?

Reply:
btw, it would be nice to avoid rewriting the file if its content is the same. We might need support from CDI for comparing specs to implement that.

Reply:
I haven't looked at the new API yet. Let's wait for https://github.com/container-orchestrated-devices/container-device-interface/pull/77 before making further changes.

Rewriting the file should happen very infrequenly. I wouldn't write extra code for it.

Reply:
Well, NodePrepareResource can be called by Kubelet any time and with any frequency. Rewriting CDI file should either be atomic(btw, spec.Write seems to do it atomically) or, better, not happening at all. We can't predict when Kubelet will  decide to call NodePrepareResource. It could happen at the same time when runtime reads previously created CDI file.

Reply:
You can use that API as it's most probably not going to change. Even if they add APIs for generating CDI file names it'll probably not change Write() API as a file path is passed to the [spec constructor](https://github.com/container-orchestrated-devices/container-device-interface/blob/main/pkg/cdi/spec.go#L93) and Write API [uses path that's set in the spec object.](https://github.com/container-orchestrated-devices/container-device-interface/blob/main/pkg/cdi/spec.go#L154)

Reply:
> Well, NodePrepareResource can be called by Kubelet any time and with any frequency. 

For the same claim? So your current code doesn't serialize that? That's okay, I'm just wondering. But even if it doesn't, the function probably will only be called when many different pods start up which all reference the same unprepared claim. That seems unlikely.

Reply:
But preferably the Cache/Registry `WriteSpec()` interface should be used instead. 

The [Spec name generation API](https://github.com/container-orchestrated-devices/container-device-interface/pull/77) that we're sorting out is designed to be used together with `WriteSpec()` / `RemoveSpec()`. `WriteSpec()` uses `Spec.Write()` under the hood...  which maybe should be an unexported Spec.write() instead.

Reply:
> So your current code doesn't serialize that?

it does. There is a lock for this purpose, so two calls can't be done at the same time. However, it doesn't guarantee that it can't be called for the (second) pod when runtime reads CDI file created by the previous call (for the firs pod referencing the same claim).

Reply:
Yes, it's unlikely to happen, I agree. But it can happen, so it's better to avoid re-generate CDI files in my opinion.

Reply:
@klihub 

> WriteSpec() uses Spec.Write() under the hood... which maybe should be an unexported Spec.write() instead.

Code in this PR does the following:
- creates spec object (passing CDI file path to it)
- marshals it to JSON
- writes json to the file (with the same CDI file path as was passed to the cdi.NewSpec)

That's why I thought that creating a spec object and then calling spec.Write() looks as a reasonable way to replace that code.

It seems not the case though. Can you explain what needs to be done instead?

Reply:
@klihub 

> But preferably the Cache/Registry WriteSpec() interface should be used instead.

Wouldn't this require reading all CDI files present in CDI directories? This can potentially slow down the plugin. However, if it's done once on the plugin start it shouldn't be a problem, right?

Reply:
I added a TODO about using newer API(s) but for now would prefer to merge the code as-is so that we can make progress with other, more important aspects (like moving the code around).